recirq/quantum_chess/circuit_transformer.py

```diff
@@ -305,6 +317,384 @@ def transform(self, circuit: cirq.Circuit) -> cirq.Circuit:
+          lg: A physical qubits graph.
+          lq: The physical qubit from which to find candidate qubits.
+        """
+        qubits = []
+        visited = set()
+        visited.add(pq)
+        queue = deque()
+        queue.append(pq)
+        while queue:
+            level = len(queue)
+            while level > 0:
+                q = queue.popleft()
+                for neighbor in pg[q]:
+                    if neighbor not in visited:
+                        visited.add(neighbor)
+                        queue.append(neighbor)
+                    if neighbor not in mapped and neighbor not in qubits:
+                        qubits.append(neighbor)
+                level -= 1
+            if len(qubits) > 0:
+                break
+        return qubits
+
+    def find_shortest_path_distance(
+        self,
+        g: Dict[cirq.GridQubit, List[cirq.GridQubit]],
+        s: cirq.GridQubit,
+        t: cirq.GridQubit,
+    ) -> int:
+        """Returns the shortest distance between the source and target qubits.
+
+        Uses breadth-first search traversal.
+
+        Args:
+          g: A physical qubits graph.
+          s: The source qubit from which to start breadth-first search.
+          t: The target qubit to search.
+        """
+        dist = defaultdict(int)
+        visited = set()
+        visited.add(s)
+        queue = deque()
+        queue.append(s)
+        while queue:
+            q = queue.popleft()
+            if q == t:
+                return dist[t]
+            for neighbor in g[q]:
+                if neighbor not in visited:
+                    dist[neighbor] = dist[q] + 1
+                    visited.add(neighbor)
+                    queue.append(neighbor)
+        return math.inf
+
+    def calculate_initial_mapping(
+        self,
+        pg: Dict[cirq.Qid, List[cirq.Qid]],
+        lg: Dict[cirq.Qid, List[Tuple[cirq.Qid, int]]],
+    ) -> Dict[cirq.Qid, cirq.GridQubit]:
+        """Returns an initial mapping of logical qubits to physical qubits.
+
+        This initial mapping algorithm is proposed by the paper "A Dynamic
+        Look-Ahead Heuristic for the Qubit Mapping Problem of NISQ Computer:
+        https://ieeexplore.ieee.org/abstract/document/8976109.
+
+        Args:
+          pg: A physical qubits graph.
+          lg: A logical qubits graph.
+        """
+        mapping = defaultdict()
+
+        pg_center = self.find_graph_center(pg)
+        lg_center = self.find_graph_center(lg)
+        mapping[lg_center] = pg_center
+
+        traversal_order = self.traverse(lg, lg_center)
+
+        while traversal_order:
+            lq = traversal_order.popleft()
+            if lq == lg_center:
+                continue
+            pq = None
+            reference_qubits = self.find_reference_qubits(mapping, lg, lq)
+            ref_q = reference_qubits[0]
+            candidate_qubits = self.find_candidate_qubits(mapping.values(), pg, ref_q)
```

Comment:
personally I found it a bit confusing with ref_q being shadowed in the loop below -- could inline reference_qubits[0] as the find_candidate_qubits() reference qubit arg instead

Reply:
updated!